internal/pkg/agent/application/actions/handlers/handler_action_diagnostics.go

```diff
@@ -0,0 +1,204 @@
+	"os"
+	"time"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/control/v2/client"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/control/v2/cproto"
+	"github.com/elastic/elastic-agent/internal/pkg/core/monitoring/config"
+	"github.com/elastic/elastic-agent/internal/pkg/diagnostics"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi/acker"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+
+	"golang.org/x/time/rate"
+)
+
+// ErrRateLimit is the rate limit error that is returned if the handler is ran too often.
+// This may occur if the user sends multiple diagnostics actions to an agent in a short duration
+// or if the agent goes offline and retrieves multiple diagnostics actions.
+// In either case the 1st action will succeed and the others will ack with an the error.
+var ErrRateLimit = fmt.Errorf("rate limit exceeded")
+
+// Uploader is the interface used to upload a diagnostics bundle to fleet-server.
+type Uploader interface {
+	UploadDiagnostics(context.Context, string, string, int64, io.Reader) (string, error)
+}
+
+// Diagnostics is the handler to process Diagnostics actions.
+// When a Diagnostics action is received a full diagnostics bundle is taken and uploaded to fleet-server.
+type Diagnostics struct {
+	log      *logger.Logger
+	coord    *coordinator.Coordinator
+	limiter  *rate.Limiter
+	uploader Uploader
+}
+
+// NewDiagnostics returns a new Diagnostics handler.
+func NewDiagnostics(log *logger.Logger, coord *coordinator.Coordinator, cfg config.Limit, uploader Uploader) *Diagnostics {
+	return &Diagnostics{
+		log:      log,
+		coord:    coord,
+		limiter:  rate.NewLimiter(rate.Every(cfg.Interval), cfg.Burst),
+		uploader: uploader,
+	}
+}
+
+// Handle processes the passed Diagnostics action.
+func (h *Diagnostics) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker) error {
+	h.log.Debugf("handlerDiagnostics: action '%+v' received", a)
+	action, ok := a.(*fleetapi.ActionDiagnostics)
+	if !ok {
+		return fmt.Errorf("invalid type, expected ActionDiagnostics and received %T", a)
+	}
+	ts := time.Now().UTC()
+	defer func() {
+		ack.Ack(ctx, action) //nolint:errcheck // no path for a failed ack
+		ack.Commit(ctx)      //nolint:errcheck //no path for failing a commit
+	}()
+
+	if !h.limiter.Allow() {
+		action.Err = ErrRateLimit
+		return ErrRateLimit
+	}
+
+	h.log.Debug("Gathering agent diagnostics.")
+	aDiag, err := h.runHooks(ctx)
+	if err != nil {
+		action.Err = err
+		return fmt.Errorf("unable to gather agent diagnostics: %w", err)
+	}
+	h.log.Debug("Gathering unit diagnostics.")
+	uDiag := h.diagUnits(ctx)
+
+	var r io.Reader
+	// attempt to create the a temporary diagnostics file on disk in order to avoid loading a
+	// potentially large file in memory.
+	// if on-disk creation fails an in-memory buffer is used.
+	f, s, err := h.diagFile(aDiag, uDiag)
```

Comment:
Using Amit's suggestion here to use an on-disk temp file and fall a back to an in-memory buffer.

Reply:
Awesome work to turn around on that feedback so quickly. Really appreciate it.